compiler/rustc_codegen_ssa/src/back/metadata.rs

```diff
@@ -258,6 +261,44 @@ pub(crate) fn create_object_file(sess: &Session) -> Option<write::Object<'static
     Some(file)
 }
 
+/// If the target specifies which platform/deployment target/SDK version
+/// should be included in the output object files, this function will
+/// return them. Apple's LD requires this to be present for Mac Catalyst.
+/// Basically LC_BUILD_VERSION
+fn macho_object_build_version_for_target(
+    target: &Target,
+) -> Option<object::write::MachOBuildVersion> {
+    let version_as_tuple = |str: &str| -> Option<(u32, u32)> {
+        str.split_once('.')
+            .and_then(|(a, b)| a.parse().and_then(|a| b.parse().map(|b| (a, b))).ok())
+    };
+    // The `object` crate demands "X.Y.Z encoded in nibbles as xxxx.yy.zz"
+    // e.g. minOS 14.0 = 0x000E0000, or SDK 16.2 = 0x00100200
+    let pack_version = |(major, minor): (u32, u32)| (major << 16) | (minor << 8);
+
+    let min_os = version_as_tuple(&target.macho_build_version_min_os)?;
+    let sdk = version_as_tuple(&target.macho_build_version_sdk)?;
+    let platform = match target.macho_build_version_platform.as_ref() {
+        "macos" => object::macho::PLATFORM_MACOS,
+        "maccatalyst" => object::macho::PLATFORM_MACCATALYST,
+        "ios" => object::macho::PLATFORM_IOS,
+        "tvos" => object::macho::PLATFORM_TVOS,
+        "watchos" => object::macho::PLATFORM_WATCHOS,
+        "bridgeos" => object::macho::PLATFORM_BRIDGEOS,
+        "driverkit" => object::macho::PLATFORM_DRIVERKIT,
+        "iossim" => object::macho::PLATFORM_IOSSIMULATOR,
+        "tvossim" => object::macho::PLATFORM_TVOSSIMULATOR,
+        "watchossim" => object::macho::PLATFORM_WATCHOSSIMULATOR,
+        _ => return None,
```

Comment:
Actually, if `sess.target.is_like_osx` == true, is that means that this return should be unreachable? Assuming if https://github.com/rust-lang/rust/pull/111384/files#r1199646361 applied.